### PR TITLE
verify the golang tarball against it's expected sha1

### LINKED ISSUE #47: SHA1 Checksum
https://golang.org/dl (which we scrape in `update.sh` already) includes a `SHA1 Checksum` column; while we're there scraping the "latest version", we should scrape that `goVERSION.src.tar.gz` checksum value and verify our downloads :+1:

### PR DESCRIPTION
**Pull request by mattrobenolt**
```
$ docker build --rm -t golang .
Sending build context to Docker daemon 7.168 kB
Step 0 : FROM buildpack-deps:jessie-scm
 ---> 607e965985c1
Step 1 : RUN apt-get update && apt-get install -y 		gcc libc6-dev make 		--no-install-recommends 	&& rm -rf /var/lib/apt/lists/*
 ---> Using cache
 ---> 0f5121dd42a6
Step 2 : ENV GOLANG_VERSION 1.5
 ---> Using cache
 ---> 848707a3bad7
Step 3 : ENV GOLANG_DOWNLOAD_URL https://golang.org/dl/go$GOLANG_VERSION.linux-amd64.tar.gz
 ---> Using cache
 ---> ad6927e0d810
Step 4 : ENV GOLANG_DOWNLOAD_SHA1 5817fa4b2252afdb02e11e8b9dc1d9173ef3bd5a
 ---> Using cache
 ---> 86a6f69b4e93
Step 5 : RUN mkdir -p /usr/src/golang 	&& curl -sSL "$GOLANG_DOWNLOAD_URL" -o golang.tar.gz 	&& echo "$GOLANG_DOWNLOAD_SHA1  golang.tar.gz" | sha1sum -c - 	&& tar -v -C /usr/local -xzf golang.tar.gz 	&& rm -r /usr/src/golang
 ---> Using cache
 ---> 747b36b4eaca
Step 6 : ENV GOPATH /go
 ---> Using cache
 ---> 37f7914c54e4
Step 7 : ENV PATH $GOPATH/bin:/usr/local/go/bin:$PATH
 ---> Using cache
 ---> 136191ef5c13
Step 8 : RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
 ---> Using cache
 ---> e592541eddd9
Step 9 : WORKDIR $GOPATH
 ---> Using cache
 ---> 5d574e30aff0
Step 10 : COPY go-wrapper /usr/local/bin/
 ---> Using cache
 ---> 64150e376c56
Successfully built 64150e376c56
$ docker run --rm -it golang go version
go version go1.5 linux/amd64
```

Fixes #47

:tada:

### CODE REVIEW

**Comment by yosifkit** [comment]
Nice!  Though I think we should add it to all the versions and modify `update.sh` to refresh the sha as well.  This will keep version bumps simple.

**Comment by mattrobenolt** [comment]
@yosifkit How do you think it should be done from within `update.sh`? Unless we go back to scraping the HTML page, but this was removed in 6ea1f29b1fe7e6b0b8eb89493ed5e06bac454654 in favor of checking the `VERSION` file from GitHub.

**Comment by yosifkit** [comment]
Since they don't have the sha's on the github page, we'll have to scrape the html again. :cry:

**Comment by mattrobenolt** [comment]
I was afraid of that. :) I’ll change it to go back to that.

-- 
Matt Robenolt
@mattrobenolt

**Comment by tianon** [comment]
I think we could still scrape latest version data from GitHub, and just scrap the SHA1 from the HTML (which should be easier given a full version instead of a partial match).

Might be worth snooping around to see if they've got those SHA1s in a file with a common name somewhere too (like `go....tar.xz.sha1`; although my hopes for that aren't very high).

File: 1.4/wheezy/Dockerfile [hunk 1]
```diff
@@ -7,17 +7,19 @@ RUN apt-get update && apt-get install -y \
...
 
 ENV GOLANG_VERSION 1.4.2
+ENV GOLANG_DOWNLOAD_URL https://golang.org/dl/go$GOLANG_VERSION.linux-amd64.tar.gz
+ENV GOLANG_DOWNLOAD_SHA1 6fede71082a2f57c13f2521ecd054b86ae77d0e5
 
-RUN curl -sSL https://golang.org/dl/go$GOLANG_VERSION.src.tar.gz \
-		| tar -v -C /usr/src -xz
+RUN mkdir -p /usr/src/golang \
+	&& curl -sSL "$GOLANG_DOWNLOAD_URL" -o golang.tar.gz \
+	&& echo "$GOLANG_DOWNLOAD_SHA1 golang.tar.gz" | sha1sum -c - \
```
Comment (mattrobenolt) [inline 1.4/wheezy/Dockerfile]: Any idea why these don't work on wheezy builds? I get:

```
sha1sum: standard input: no properly formatted SHA1 checksum lines found
```

Comment (mattrobenolt) [inline 1.4/wheezy/Dockerfile]: NEVERMIND, turns out, it requires literally two spaces between `<checksum>  <file>`. Not just 1. Silly me.

**Comment by mattrobenolt** [comment]
Ugh, so it turns out that we need a `src` package for 1.3 and 1.4 versions only, while 1.5 can support a binary package.

**Comment by mattrobenolt** [comment]
@yosifkit @tianon Everything looks good now, and all images build.

**Comment by mattrobenolt** [comment]
Going to rebase this now for 1.5 release. :) Would be nice to get this merged.

File: update.sh [hunk 2]
```diff
@@ -11,21 +11,31 @@ versions=( "${versions[@]%/}" )
...
+	# while 1.4 installs from src
+	if [ "$version" = '1.4' ]; then
+		package='src'
 	fi
+	fullVersion="$(echo $source | grep -Po '">go'"$version"'.*?\.'"$package"'\.tar\.gz</a>' | sed -r 's!.*go([^"/<]+)\.'"$package"'\.tar\.gz.*!\1!' | sort -V | tail -1)"
 	if [ -z "$fullVersion" ]; then
 		echo >&2 "warning: cannot find full version for $version"
 		continue
 	fi
+	checksum="$(echo $source | grep -Po '">go'"$fullVersion"'\.'"$package"'\.tar\.gz</a>.*?>[a-f0-9]{40}<' | sed -r 's!.*([a-f0-9]{40}).*!\1!' | sort -V | tail -1)"
```
Comment (yosifkit) [inline update.sh]: Shouldn't this `sort` be removed?

Comment (mattrobenolt) [inline update.sh]: Oops, yeah. Fortunately, it doesn't change the output in this case. :)

**Comment by yosifkit** [comment]
LGTM

**Comment by tianon** [comment]
I think we should still scrape latest version data from GitHub, and just scrape the SHA1 from the HTML, since it'll be more accurate than the HTML scraping.  WDYT?

**Comment by mattrobenolt** [comment]
@tianon, yeah, I can bring that back. Just assumed that since we were going back to scraping the Google page, we could solely use that, but your reasoning makes sense.

**Comment by mattrobenolt** [comment]
@tianon Added back.

File: 1.5/Dockerfile [hunk 3]
```diff
@@ -6,12 +6,15 @@ RUN apt-get update && apt-get install -y \
...
+ENV GOLANG_DOWNLOAD_URL https://golang.org/dl/go$GOLANG_VERSION.linux-amd64.tar.gz
+ENV GOLANG_DOWNLOAD_SHA1 5817fa4b2252afdb02e11e8b9dc1d9173ef3bd5a
 
-RUN curl -sSL https://golang.org/dl/go$GOLANG_VERSION.$GOLANG_GOOS-$GOLANG_GOARCH.tar.gz \
-		| tar -v -C /usr/local -xz
+RUN mkdir -p /usr/src/golang \
+	&& curl -sSL "$GOLANG_DOWNLOAD_URL" -o golang.tar.gz \
+	&& echo "$GOLANG_DOWNLOAD_SHA1  golang.tar.gz" | sha1sum -c - \
+	&& tar -v -C /usr/local -xzf golang.tar.gz \
+	&& rm -r /usr/src/golang
```
Comment (tianon) [inline 1.5/Dockerfile]: Is `/usr/src/golang` here used for anything?

(while you're there, mind adding `-f` to the `curl` usage and removing `-v` from `tar`? :innocent:)

**Comment by tianon** [comment]
Ok, that's my final gripe.  This looks solid, thanks! :heart:

File: 1.5/Dockerfile (hunk 3 above)
Comment (mattrobenolt) [inline 1.5/Dockerfile]: > Is /usr/src/golang here used for anything?

No, that's why it's being removed. :) It's just to hold the source file to download, then gets removed after.

> (while you're there, mind adding -f to the curl usage and removing -v from tar? :innocent:)

Sure.

**Comment by mattrobenolt** [comment]
@tianon done.

File: 1.5/Dockerfile [hunk 4]
```diff
@@ -6,12 +6,15 @@ RUN apt-get update && apt-get install -y \
...
-ENV GOLANG_GOOS linux
-ENV GOLANG_GOARCH amd64
 ENV GOLANG_VERSION 1.5
+ENV GOLANG_DOWNLOAD_URL https://golang.org/dl/go$GOLANG_VERSION.linux-amd64.tar.gz
+ENV GOLANG_DOWNLOAD_SHA1 5817fa4b2252afdb02e11e8b9dc1d9173ef3bd5a
 
-RUN curl -sSL https://golang.org/dl/go$GOLANG_VERSION.$GOLANG_GOOS-$GOLANG_GOARCH.tar.gz \
-		| tar -v -C /usr/local -xz
+RUN mkdir -p /usr/src/golang \
+	&& curl -fsSL "$GOLANG_DOWNLOAD_URL" -o golang.tar.gz \
```
Comment (yosifkit) [inline 1.5/Dockerfile]: Um, this just downloads to `/`, so the `mkdir` and `rm` don't actually do anything.

Comment (yosifkit) [inline 1.5/Dockerfile]: We could just keep the download to `/` and then `rm` the tar at the end.

Comment (mattrobenolt) [inline 1.5/Dockerfile]: Yeah, good call. Not sure why I added it, but it's removed now.

**Comment by yosifkit** [comment]
LGTM, really this time.

**Comment by tianon** [comment]
LGTM :+1: :heart:

**Pull request merged**